gnrc_ipv6_nib/SLAAC: rfc7217 stable privacy addresses

xnumad opened the pull request:
### Contribution description

Implementation of RFC7217 ("stable privacy") means that for addresses generated by SLAAC, their IID (interface identifier) is
- semantically opaque, i.e. cannot derive hwaddr from it
- random but with such parameters that it is stable within a subnet (i.e. for the same prefix on the same interface)
- derived from a randomly generated key (secret_key) that shouldn't be shared across devices.
RIOT-specific: The implementation in this PR compiles the secret_key into the elffile, therefore the same elffile shouldn't be flashed onto multiple devices - in order to fulfill the RFC requirements.

Notes about implementation:
- uses tail recursion (i.e. optimizable by compiler, or manually replacable by a loop) (affects stack size)

RFC compatibility: Requirement levels: (implemented: MUST, SHOULD. not implemented: MAY, OPTIONAL.)

--

#### Adaptability

Supports any link layer that has a hardware address:
https://github.com/xnumad/RIOT/blob/88363f3cde7175691fe27399d73a54e08934e6c7/sys/net/gnrc/network_layer/ipv6/nib/_nib-slaac.c#L168-L169

The usage of IIDs which do not match the link layer address causes LOWPAN_IPHC to not be able to statelessly compress the IP address anymore. An optimization to enable compression again could be to add compression contexts (feature branch for opportunistic compression contexts: https://github.com/xnumad/RIOT/tree/feature%2Fopportunistic-compression-contexts).

##### 6LoWPAN specifics
They do not affect this PRs current functionality but are worth noting.

For link-local addresses on a 6LN iface, this PR does not use the IDGEN (interface identifier generation) mechanism described by rfc7217.
- This is because of incompatibility with 6LoWPAN-ND (RFC6775) (and RFC8505, which would allow it, is not implemented).
- → Not fulfilling RFC7217 requirement "MUST be employed for [...] link-local"

### Testing procedure

Add `CFLAGS += -DCONFIG_GNRC_IPV6_STABLE_PRIVACY=1` at the appropriate position in the Makefile of `examples/gnrc_networking`. Tested on BOARD=nrf52840dk

Output of `ifconfig` command shows that SLAAC addresses have random differing IIDs (except for link-local address on a 6LoWPAN iface, see above). The IID is stable within a subnet, i.e. it also persists across reboots.

### Issues/PRs references

Commits marked with 🍒 (cherry-picked) are largely derived from the RFC8981 implementation at #20369

xnumad committed:
config: Define feature flag

xnumad committed:
config: RFC params

xnumad committed:
includes

xnumad committed:
IDGEN

xnumad committed:
Retry condition check

xnumad committed:
Overload method

xnumad committed:
Generate secret_key at compilation

Requires Python >=3.6, which is already EOL, so not noteworthy
https://github.com/RIOT-OS/RIOT/pull/20370#discussion_r1498370589

xnumad committed:
Use IDGEN

xnumad committed:
Handle DAD

xnumad committed:
Fix `unused-function`

xnumad committed:
IDGEN error handling and logging

xnumad committed:
Perform DAD for RPL PIO

xnumad committed:
Allow rfc7217 IDGEN without L2 addr

Move pre-condition check to the corresponding method requiring this condition

xnumad committed:
RFC7217 for NETOPT_IPV6_IID

xnumad committed:
Rename method

xnumad committed:
Rename method

xnumad committed:
No RFC7217 for 6LN link-local addresses

Wrap in `if(is_rfc7217)`, use normal IDGEN if false

xnumad committed:
IDGEN: require L2 addr

Remove fallback to netif.pid because presumably doesn't fulfill the RFC stability requirements of "constant across system bootstrap sequences and other network events (e.g., bringing another interface up or down)"

xnumad committed:
finalize: vera++ style check

xnumad committed:
Wrap retries flag in IS_ACTIVE

xnumad committed:
🍒SLAAC only for /64 prefixes

xnumad committed:
🍒Supress SLAAC trigger

Need to check for matching address rather than prefix because for a 6LN, the SLAAC prefix is never stored in the prefix list, since everything is considered off-link.

xnumad committed:
🍒gnrc_netif_ipv6_addr_pfx_idx

xnumad committed:
🍒retries: New flag

xnumad committed:
Wrap IANA IID check in IS_ACTIVE

xnumad committed:
🍒Documentation

xnumad committed:
🍒creation: Check IANA IID

xnumad committed:
🍒Documentation for constant

xnumad committed:
Wrap

xnumad committed:
doccheck

chrysn commented on sys/net/gnrc/Makefile.dep:
```diff
@@ -441,6 +441,25 @@ ifneq (,$(filter gnrc_ipv6_nib,$(USEMODULE)))
+
+  ##prepare value
+  stable_privacy_secret_key != python3 -c "import secrets; print(','.join([f'0x{byte:02x}' for byte in secrets.token_bytes(16)]))"
+  #Python >= 3.6, because:
+  #- 3.6 https://docs.python.org/3/library/secrets.html
+  #- 3.6 https://docs.python.org/3/reference/lexical_analysis.html#f-strings
+    #replaceable: f'0x{byte:02x}' == '0x' + format(byte, '02x')
```
Good you checked, but given 3.7 is already EoL for good, the comments are probably unwarranted.

chrysn commented on sys/net/gnrc/Makefile.dep:
```diff
@@ -441,6 +441,25 @@ ifneq (,$(filter gnrc_ipv6_nib,$(USEMODULE)))
+  # Set another macro that is needed for this option.
+
+  ##prepare value
+  stable_privacy_secret_key != python3 -c "import secrets; print(','.join([f'0x{byte:02x}' for byte in secrets.token_bytes(16)]))"
```
We'll have to think a bit about device individualization in RIOT as a whole (esp. for questions like "is this maybe better created in a HWRNG assisted way at first start?"), but until then, passing it into the build seems fine.

As someone who repeatedly flashes devices and then starts network interactions, it may be convenient to store this in the build directory and recreate only if absent, as that will make addresses stable across rebuilds.

xnumad replied:
Definitely agree, letting the board generate the secret_key by itself (and permanently store it - as per the RFC) would be more convenient, as a future improvement!

chrysn commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-slaac.c:
```diff
@@ -30,8 +36,20 @@
 void _auto_configure_addr(gnrc_netif_t *netif, const ipv6_addr_t *pfx,
                           uint8_t pfx_len)
+#else
+inline void _auto_configure_addr(gnrc_netif_t *netif, const ipv6_addr_t *pfx,
```
Having this either-declared-or-inline is quite confusing, especially being split across two files; do you see options to simplify that?

xnumad replied:
This either-declared-or-inline is to achieve a logic for method overloading.
If it's confusing, one could simply remove the `inline` qualifier 😄 (at slight cost of performance)

(Additionally, to revert this slight performance penalty,
one could move the logic for the method overload to each caller.
Like some existing code does (method has one caller only): https://github.com/RIOT-OS/RIOT/blob/32d1cd93a169335e5629f63d55ccc11406009681/sys/net/gnrc/network_layer/ipv6/nib/nib.c#L808-L816
With a definition like:
https://github.com/RIOT-OS/RIOT/blob/32d1cd93a169335e5629f63d55ccc11406009681/sys/net/gnrc/network_layer/ipv6/nib/nib.c#L1610-L1617

This would affect 3 existing callers. Instead of having it in one place to cover all callers at once, this puts the burden on the caller however, which may cause incomplete feature coverage if overlooked when adding a new caller.)

chrysn replied:
I don't have the full picture, but if as I understand this should only affect callers in this compilation unit, would it work (i.e. maintain all the relevant properties) to have something like this?

```c
static inline void _our_auto_configure_addr(gnrc_netif_t *netif, const ipv6_addr_t *pfx,
                          uint8_t pfx_len)
{
#if !IS_ACTIVE(CONFIG_GNRC_IPV6_STABLE_PRIVACY)
    _auto_configure_addr(netif, pfx, len);
#else
    _auto_configure_addr_with_dad_ctr(netif, pfx, pfx_len, 0);
#endif
}
```

chrysn commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-slaac.c:
```diff
@@ -96,7 +138,91 @@ void _auto_configure_addr(gnrc_netif_t *netif, const ipv6_addr_t *pfx,
+}
+
+inline bool _stable_privacy_should_retry_idgen(uint8_t *dad_ctr, const char *reason) {
+    if (*dad_ctr < STABLE_PRIVACY_IDGEN_RETRIES) { //within retry limit
```
At the very last stages of review, `//` style comments will be flagged for code style. Sorry it's not automated or you'd have seen it in CI :-/

chrysn reviewed (commented):
I don't have the time to do a full review at the moment, but from a quick high-level reading this looks like it's well done, and checks a lot of boxes (including careful treatment of Kconfig handling).

A few notes around here and there are really just "things that struck me" with no particular severity. Given there are a lot of fixup commits from the onset, please consider squashing (possibly after handling any of my comments), as that'll make the sequence of commits easier to understand.

xnumad committed:
chore: Use C-style comments

https://github.com/RIOT-OS/RIOT/pull/20370#discussion_r1498378096 and CODING_CONVENTIONS.md

xnumad commented:
Squashed

xnumad committed:
refactor: Make called function inline

xnumad committed:
refactor: Rename

_auto_configure_addr -> auto_configure_addr
_auto_configure_addr_default -> _auto_configure_addr

xnumad commented:
Squashed a typo

xnumad committed:
fix: Don't generate already assigned addresses

Not only for the caller `_auto_configure_addr`. Other caller is NETOPT_IPV6_IID. Move check to callee.
Probably low severity because address conflicts are unlikely anyway.

xnumad committed:
fixup! RFC7217 for NETOPT_IPV6_IID

`-Werror=maybe-uninitialized`

xnumad committed:
fix: Idempotency wrapper for callers

Example for a caller assuming idempotency: uhcp (-> `gnrc_netif_ipv6_add_prefix` -> `NETOPT_IPV6_IID`)

xnumad commented:
Rebased.

xnumad commented:
Rebased to resolve merge conflicts

crasbe commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-slaac.c:
```diff
@@ -96,7 +130,120 @@ void _auto_configure_addr(gnrc_netif_t *netif, const ipv6_addr_t *pfx,
+#endif
+
 #if IS_ACTIVE(CONFIG_GNRC_IPV6_NIB_SLAAC)
+#if !IS_ACTIVE(CONFIG_GNRC_IPV6_STABLE_PRIVACY)
```
Also the following commands should get indentations. I don't want to add too many review comments, but you get the idea.

crasbe commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-slaac.h:
```diff
@@ -40,12 +89,75 @@ extern "C" {
+#if IS_ACTIVE(CONFIG_GNRC_IPV6_STABLE_PRIVACY)
+/**
+ * @brief Overload of @ref _auto_configure_addr
+ * @param dad_ctr rfc7217 DAD_Counter
```
Although it would probaly be good to document the other parameters too, as well as in the following function.

crasbe commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-slaac.h:
```diff
@@ -40,12 +89,75 @@ extern "C" {
+ * It is assumed to be of length @ref SLAAC_PREFIX_LENGTH
+ * @param[in,out] dad_ctr ("DAD_Counter" in rfc7217)
+ * Value may increase and is to be stored associated with the address by caller
+ * @return 0 on success
+ * @return -1 if failed, because retry limit reached
+ * @return  `-ENOTSUP`, if interface has no link-layer address.
```
```suggestion
 * @retval 0 on success
 * @retval -1 if failed, because retry limit reached
 * @retval  `-ENOTSUP`, if interface has no link-layer address.
```

crasbe commented on sys/net/gnrc/network_layer/ipv6/nib/_nib-slaac.h:
```diff
@@ -40,12 +89,75 @@ extern "C" {
+ */
+bool _iid_is_iana_reserved(const eui64_t *iid);
+
+/**
+ * @param[in,out] dad_ctr
+ * @param[in] reason
+ */
```
This documentation is still incomplete.

crasbe reviewed (commented):
As with the other review, I can't really comment on the content of the changes.

carl-tud commented:
I've noticed this PR would only let users either fully enable or fully disable 7217 RIDs. However, perhaps it's desirable to enable 7217 more selectively. For example, [9428 6lo NFC (niche, I'll admit)](https://datatracker.ietf.org/doc/html/rfc9428) _always_ requires 7217, but maybe I don't want 7217 for other link layer protocols / `netif`s. Maybe someone with more in-depth knowledge could chime in. 

Suggestion `GNRC_NETIF_FLAGS_IPV6_STABLE_PRIVACY` in `sys/include/net/gnrc/netif/flags.h

carl-tud commented:
Also, I think header function decls and struct decls shouldn't be gated behind a config/module macro guard. 1, we might see compilation errors only when that module is imported and, 2, there's no overhead from having decls. After all, it's the module impl (compilation units aka. `.c` files) that will be conditionally included if imported.

edit: For reference, we had this discussion in the RIOT chat recently. you might want to permit writing constructs like this:
```c
void main(void) {
  if (IS_USED(MODULE_FOO_BAR)) {
    ipv6_get_rfc7217_iid(...);
  }
}
```

this is supposed to always compile, but wouldn't under this PR.
might be a general issue in RIOT, I've just realised.

edit, vol. II: we should provide guidance in CONTRIBUTING.md on what should be hidden behind a `CONFIG_` flag and what should be a pseudomodule. difference in code is just `IS_ACTIVE(CONFIG_...)` vs `IS_USED(MODULE_...)`.

carl-tud commented:
I don't know if this is intended, but `gnrc_netif_ipv6_iid_from_addr`/`l2util_ipv6_iid_from_addr` etc. still returns the eui64 generated from the L2 addr, so in a lot of places, this PR does not work because these functions do not accept the prefix as an argument, hence cannot calculate the 7217 IID.

It's also unclear to me how these `gnrc_netif_ipv6_iid_to_addr` APIs in RIOT are supposed to map a 7217 IID back to an L2 addr... All of this is more critiquing our API design as opposed to this PR, I guess...
